File: templates/langchain/rubric/{{cookiecutter.project_slug}}/frontend/src/contracts.ts
export const TASK_PROMPT =
  "Implement find_duplicates(values). Return valid Python source only.";

export const BASELINE_RUBRIC = `- The response is valid Python source that defines find_duplicates(values).
- Each duplicated value appears exactly once in the returned list.
- Result order follows when each value first becomes a duplicate.
- Unhashable values such as nested lists are supported.
- The input sequence is not mutated.
- Before satisfied, call run_test_suite with the exact current candidate and receive ok=true.
`;

export const DEFAULT_MAX_ITERATIONS = 3;
export const MAX_ITERATIONS = 20;

export type RunMode = "demo" | "live";
export type RubricResult =
  | "satisfied"
  | "needs_revision"
  | "max_iterations_reached"
  | "failed"
  | "grader_error";
export type TerminalStatus = Exclude<RubricResult, "needs_revision">;
export type GateReason =
  | "satisfied_with_current_evidence"
  | "terminal_status_not_satisfied"
  | "current_evidence_missing";

export type WireUIEvent = {
  event_id: string;
  type: string;
  grading_run_id: string;
  iteration: number;
  candidate_version: number | null;
  candidate_id: string | null;
  payload: unknown;
};

type EventBase = {
  eventId: string;
  gradingRunId: string;
  iteration: number;
};

export type CandidateEvent = EventBase & {
  type: "candidate";
  candidateVersion: number;
  candidateId: string;
  source: string | null;
  sourceOmitted?: boolean;
};

export type GraderFeedbackEvent = EventBase & {
  type: "grader_feedback";
  candidateVersion: number | null;
  candidateId: string | null;
  message: string;
};

export type EvidenceResult = {
  requestedCandidateId: string;
  ok: boolean;
  behaviorFailures: string[];
  profileFailures: string[];
  durationMs: number;
  timedOut: boolean;
  outputTruncated: boolean;
};

export type RubricEvidenceEvent = EventBase & {
  type: "rubric_evidence";
  candidateVersion: number;
  candidateId: string;
  evidence: EvidenceResult;
};

export type RubricEvaluationStartEvent = EventBase & {
  type: "rubric_evaluation_start";
  candidateVersion: number | null;
  candidateId: string | null;
};

export type CriterionResult = {
  criterion: string;
  passed: boolean;
  gap: string;
};

export type RubricEvaluationEndEvent = EventBase & {
  type: "rubric_evaluation_end";
  candidateVersion: number | null;
  candidateId: string | null;
  result: RubricResult;
  explanation: string;
  criteria: CriterionResult[];
};

export type UIEvent =
  | CandidateEvent
  | GraderFeedbackEvent
  | RubricEvidenceEvent
  | RubricEvaluationStartEvent
  | RubricEvaluationEndEvent;

export type CandidateRecord = {
  gradingRunId: string;
  version: number;
  iteration: number;
  candidateId: string;
  source: string | null;
  sourceOmitted?: boolean;
};

export type EvidenceRecord = EvidenceResult & {
  eventId: string;
  gradingRunId: string;
  iteration: number;
  candidateVersion: number;
  candidateId: string;
};

export type EvaluationRecord = {
  eventId: string;
  gradingRunId: string;
  iteration: number;
  candidateVersion: number | null;
  candidateId: string | null;
  result: RubricResult;
  explanation: string;
  criteria: CriterionResult[];
};

export type FeedbackRecord = {
  eventId: string;
  gradingRunId: string;
  iteration: number;
  candidateVersion: number | null;
  candidateId: string | null;
  message: string;
};

export type RunReport = {
  mode: RunMode;
  threadId: string;
  innerThreadId: string;
  gradingRunId: string;
  terminalStatus: TerminalStatus;
  accepted: boolean;
  gateReason: GateReason;
  iterations: number;
  candidates: CandidateRecord[];
  finalCandidate: string | null;
  evidence: EvidenceRecord[];
  evaluations: EvaluationRecord[];
  feedback: FeedbackRecord[];
};

export type ReportsByMode = Record<RunMode, RunReport | null>;

const RUBRIC_RESULTS = new Set<RubricResult>([
  "satisfied",
  "needs_revision",
  "max_iterations_reached",
  "failed",
  "grader_error",
]);
const TERMINAL_STATUSES = new Set<TerminalStatus>([
  "satisfied",
  "max_iterations_reached",
  "failed",
  "grader_error",
]);
const GATE_REASONS = new Set<GateReason>([
  "satisfied_with_current_evidence",
  "terminal_status_not_satisfied",
  "current_evidence_missing",
]);
const CANDIDATE_ID = /^[0-9a-f]{64}$/;

function isObject(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

function isNonEmptyString(value: unknown): value is string {
  return typeof value === "string" && value.length > 0;
}

function isCandidateId(value: unknown): value is string {
  return typeof value === "string" && CANDIDATE_ID.test(value);
}

function candidateSource(
  value: unknown,
  sourceOmitted: boolean,
): string | null | undefined {
  if (sourceOmitted) return value === null ? null : undefined;
  return typeof value === "string" ? value : undefined;
}

function isNonNegativeInteger(value: unknown): value is number {
  return Number.isInteger(value) && (value as number) >= 0;
}

function isPositiveInteger(value: unknown): value is number {
  return Number.isInteger(value) && (value as number) >= 1;
}

function stringArray(value: unknown): string[] | null {
  if (!Array.isArray(value) || !value.every((item) => typeof item === "string")) {
    return null;
  }
  return [...value];
}

function candidateBinding(
  version: unknown,
  candidateId: unknown,
): { candidateVersion: number | null; candidateId: string | null } | null {
  if (version === null && candidateId === null) {
    return { candidateVersion: null, candidateId: null };
  }
  if (!isPositiveInteger(version) || !isCandidateId(candidateId)) {
    return null;
  }
  return { candidateVersion: version, candidateId };
}

function criterion(value: unknown): CriterionResult | null {
  if (
    !isObject(value) ||
    typeof value.criterion !== "string" ||
    typeof value.passed !== "boolean" ||
    typeof value.gap !== "string"
  ) {
    return null;
  }
  return {
    criterion: value.criterion,
    passed: value.passed,
    gap: value.gap,
  };
}

function criteria(value: unknown): CriterionResult[] | null {
  if (!Array.isArray(value)) return null;
  const decoded = value.map(criterion);
  return decoded.every((item): item is CriterionResult => item !== null)
    ? decoded
    : null;
}

function evidenceResult(value: unknown): EvidenceResult | null {
  if (!isObject(value)) return null;
  const behaviorFailures = stringArray(value.behavior_failures);
  const profileFailures = stringArray(value.profile_failures);
  if (
    !isCandidateId(value.requested_candidate_id) ||
    typeof value.ok !== "boolean" ||
    behaviorFailures === null ||
    profileFailures === null ||
    !isNonNegativeInteger(value.duration_ms) ||
    typeof value.timed_out !== "boolean" ||
    typeof value.output_truncated !== "boolean"
  ) {
    return null;
  }
  return {
    requestedCandidateId: value.requested_candidate_id,
    ok: value.ok,
    behaviorFailures,
    profileFailures,
    durationMs: value.duration_ms,
    timedOut: value.timed_out,
    outputTruncated: value.output_truncated,
  };
}

function evidenceMatchesCandidate(
  evidence: EvidenceResult,
  candidateId: string,
): boolean {
  return (
    evidence.requestedCandidateId === candidateId ||
    (!evidence.ok && evidence.profileFailures.includes("candidate_binding"))
  );
}

function eventEnvelope(value: unknown): {
  wire: WireUIEvent;
  binding: { candidateVersion: number | null; candidateId: string | null };
} | null {
  if (
    !isObject(value) ||
    !isNonEmptyString(value.event_id) ||
    typeof value.type !== "string" ||
    !isNonEmptyString(value.grading_run_id) ||
    !isNonNegativeInteger(value.iteration)
  ) {
    return null;
  }
  const binding = candidateBinding(value.candidate_version, value.candidate_id);
  if (binding === null) return null;
  return {
    wire: {
      event_id: value.event_id,
      type: value.type,
      grading_run_id: value.grading_run_id,
      iteration: value.iteration,
      candidate_version: binding.candidateVersion,
      candidate_id: binding.candidateId,
      payload: value.payload,
    },
    binding,
  };
}

export function decodeUIEvent(value: unknown): UIEvent | null {
  const envelope = eventEnvelope(value);
  if (envelope === null) return null;
  const { wire, binding } = envelope;
  const payload = wire.payload;
  if (!isObject(payload)) return null;
  const base = {
    eventId: wire.event_id,
    gradingRunId: wire.grading_run_id,
    iteration: wire.iteration,
  };

  switch (wire.type) {
    case "candidate":
      if (
        payload.source_omitted !== undefined &&
        typeof payload.source_omitted !== "boolean"
      ) {
        return null;
      }
      const sourceOmitted = payload.source_omitted === true;
      const source = candidateSource(payload.source, sourceOmitted);
      if (
        binding.candidateVersion === null ||
        binding.candidateId === null ||
        source === undefined
      ) {
        return null;
      }
      return {
        ...base,
        type: "candidate",
        candidateVersion: binding.candidateVersion,
        candidateId: binding.candidateId,
        source,
        sourceOmitted,
      };
    case "grader_feedback":
      if (typeof payload.message !== "string") return null;
      return {
        ...base,
        type: "grader_feedback",
        ...binding,
        message: payload.message,
      };
    case "rubric_evidence": {
      const evidence = evidenceResult(payload);
      if (
        evidence === null ||
        binding.candidateVersion === null ||
        binding.candidateId === null ||
        !evidenceMatchesCandidate(evidence, binding.candidateId)
      ) {
        return null;
      }
      return {
        ...base,
        type: "rubric_evidence",
        candidateVersion: binding.candidateVersion,
        candidateId: binding.candidateId,
        evidence,
      };
    }
    case "rubric_evaluation_start":
      return {
        ...base,
        type: "rubric_evaluation_start",
        ...binding,
      };
    case "rubric_evaluation_end": {
      const decodedCriteria = criteria(payload.criteria);
      if (
        !RUBRIC_RESULTS.has(payload.result as RubricResult) ||
        typeof payload.explanation !== "string" ||
        decodedCriteria === null
      ) {
        return null;
      }
      return {
        ...base,
        type: "rubric_evaluation_end",
        ...binding,
        result: payload.result as RubricResult,
        explanation: payload.explanation,
        criteria: decodedCriteria,
      };
    }
    default:
      return null;
  }
}

function candidateRecord(value: unknown): CandidateRecord | null {
  if (
    !isObject(value) ||
    (value.source_omitted !== undefined &&
      typeof value.source_omitted !== "boolean")
  ) {
    return null;
  }
  const sourceOmitted = value.source_omitted === true;
  const source = candidateSource(value.source, sourceOmitted);
  if (
    !isNonEmptyString(value.grading_run_id) ||
    !isPositiveInteger(value.version) ||
    !isNonNegativeInteger(value.iteration) ||
    !isCandidateId(value.candidate_id) ||
    source === undefined
  ) {
    return null;
  }
  return {
    gradingRunId: value.grading_run_id,
    version: value.version,
    iteration: value.iteration,
    candidateId: value.candidate_id,
    source,
    sourceOmitted,
  };
}

function evidenceRecord(value: unknown): EvidenceRecord | null {
  if (
    !isObject(value) ||
    !isNonEmptyString(value.event_id) ||
    !isNonEmptyString(value.grading_run_id) ||
    !isNonNegativeInteger(value.iteration) ||
    !isPositiveInteger(value.candidate_version) ||
    !isCandidateId(value.candidate_id)
  ) {
    return null;
  }
  const decodedEvidence = evidenceResult(value);
  if (decodedEvidence === null) return null;
  return {
    eventId: value.event_id,
    gradingRunId: value.grading_run_id,
    iteration: value.iteration,
    candidateVersion: value.candidate_version,
    candidateId: value.candidate_id,
    ...decodedEvidence,
  };
}

function evaluationRecord(value: unknown): EvaluationRecord | null {
  if (
    !isObject(value) ||
    !isNonEmptyString(value.event_id) ||
    !isNonEmptyString(value.grading_run_id) ||
    !isNonNegativeInteger(value.iteration) ||
    !RUBRIC_RESULTS.has(value.result as RubricResult) ||
    typeof value.explanation !== "string"
  ) {
    return null;
  }
  const binding = candidateBinding(value.candidate_version, value.candidate_id);
  const decodedCriteria = criteria(value.criteria);
  if (binding === null || decodedCriteria === null) return null;
  return {
    eventId: value.event_id,
    gradingRunId: value.grading_run_id,
    iteration: value.iteration,
    ...binding,
    result: value.result as RubricResult,
    explanation: value.explanation,
    criteria: decodedCriteria,
  };
}

function feedbackRecord(value: unknown): FeedbackRecord | null {
  if (
    !isObject(value) ||
    !isNonEmptyString(value.event_id) ||
    !isNonEmptyString(value.grading_run_id) ||
    !isNonNegativeInteger(value.iteration) ||
    typeof value.message !== "string"
  ) {
    return null;
  }
  const binding = candidateBinding(value.candidate_version, value.candidate_id);
  if (binding === null) return null;
  return {
    eventId: value.event_id,
    gradingRunId: value.grading_run_id,
    iteration: value.iteration,
    ...binding,
    message: value.message,
  };
}

function decodeArray<T>(
  value: unknown,
  decoder: (item: unknown) => T | null,
): T[] | null {
  if (!Array.isArray(value)) return null;
  const decoded = value.map(decoder);
  return decoded.every((item): item is T => item !== null) ? decoded : null;
}

function recordsMatchCandidates(report: RunReport): boolean {
  const versions = report.candidates.map((candidate) => candidate.version);
  if (new Set(versions).size !== versions.length) return false;
  const eventIds = [
    ...report.evidence.map((item) => item.eventId),
    ...report.evaluations.map((item) => item.eventId),
    ...report.feedback.map((item) => item.eventId),
  ];
  if (new Set(eventIds).size !== eventIds.length) return false;
  const candidates = new Map(
    report.candidates.map((candidate) => [candidate.version, candidate.candidateId]),
  );
  if (
    report.candidates.some(
      (candidate) => candidate.gradingRunId !== report.gradingRunId,
    ) ||
    report.evidence.some(
      (item) =>
        item.gradingRunId !== report.gradingRunId ||
        !evidenceMatchesCandidate(item, item.candidateId) ||
        candidates.get(item.candidateVersion) !== item.candidateId,
    )
  ) {
    return false;
  }
  return [...report.evaluations, ...report.feedback].every(
    (item) =>
      item.gradingRunId === report.gradingRunId &&
      (item.candidateVersion === null ||
        candidates.get(item.candidateVersion) === item.candidateId),
  );
}

function decodeWireRunReport(value: Record<string, unknown>): RunReport | null {
  if (
    (value.mode !== "demo" && value.mode !== "live") ||
    !isNonEmptyString(value.thread_id) ||
    !isNonEmptyString(value.inner_thread_id) ||
    !isNonEmptyString(value.grading_run_id) ||
    !TERMINAL_STATUSES.has(value.terminal_status as TerminalStatus) ||
    typeof value.accepted !== "boolean" ||
    !GATE_REASONS.has(value.gate_reason as GateReason) ||
    !isNonNegativeInteger(value.iterations) ||
    (value.final_candidate !== null && typeof value.final_candidate !== "string")
  ) {
    return null;
  }
  const candidates = decodeArray(value.candidates, candidateRecord);
  const evidence = decodeArray(value.evidence, evidenceRecord);
  const evaluations = decodeArray(value.evaluations, evaluationRecord);
  const feedback = decodeArray(value.feedback, feedbackRecord);
  if (
    candidates === null ||
    evidence === null ||
    evaluations === null ||
    feedback === null
  ) {
    return null;
  }
  const report: RunReport = {
    mode: value.mode,
    threadId: value.thread_id,
    innerThreadId: value.inner_thread_id,
    gradingRunId: value.grading_run_id,
    terminalStatus: value.terminal_status as TerminalStatus,
    accepted: value.accepted,
    gateReason: value.gate_reason as GateReason,
    iterations: value.iterations,
    candidates,
    finalCandidate: value.final_candidate as string | null,
    evidence,
    evaluations,
    feedback,
  };
  return recordsMatchCandidates(report) ? report : null;
}

function normalizedReportToWire(value: Record<string, unknown>): Record<string, unknown> {
  const candidates = Array.isArray(value.candidates)
    ? value.candidates.map((item) =>
        isObject(item)
          ? {
              grading_run_id: item.gradingRunId,
              version: item.version,
              iteration: item.iteration,
              candidate_id: item.candidateId,
              source: item.source,
              source_omitted: item.sourceOmitted ?? false,
            }
          : item,
      )
    : value.candidates;
  const evidence = Array.isArray(value.evidence)
    ? value.evidence.map((item) =>
        isObject(item)
          ? {
              event_id: item.eventId,
              grading_run_id: item.gradingRunId,
              iteration: item.iteration,
              candidate_version: item.candidateVersion,
              candidate_id: item.candidateId,
              requested_candidate_id: item.requestedCandidateId,
              ok: item.ok,
              behavior_failures: item.behaviorFailures,
              profile_failures: item.profileFailures,
              duration_ms: item.durationMs,
              timed_out: item.timedOut,
              output_truncated: item.outputTruncated,
            }
          : item,
      )
    : value.evidence;
  const evaluations = Array.isArray(value.evaluations)
    ? value.evaluations.map((item) =>
        isObject(item)
          ? {
              event_id: item.eventId,
              grading_run_id: item.gradingRunId,
              iteration: item.iteration,
              candidate_version: item.candidateVersion,
              candidate_id: item.candidateId,
              result: item.result,
              explanation: item.explanation,
              criteria: item.criteria,
            }
          : item,
      )
    : value.evaluations;
  const feedback = Array.isArray(value.feedback)
    ? value.feedback.map((item) =>
        isObject(item)
          ? {
              event_id: item.eventId,
              grading_run_id: item.gradingRunId,
              iteration: item.iteration,
              candidate_version: item.candidateVersion,
              candidate_id: item.candidateId,
              message: item.message,
            }
          : item,
      )
    : value.feedback;
  return {
    mode: value.mode,
    thread_id: value.threadId,
    inner_thread_id: value.innerThreadId,
    grading_run_id: value.gradingRunId,
    terminal_status: value.terminalStatus,
    accepted: value.accepted,
    gate_reason: value.gateReason,
    iterations: value.iterations,
    candidates,
    final_candidate: value.finalCandidate,
    evidence,
    evaluations,
    feedback,
  };
}

export function decodeRunReport(value: unknown): RunReport | null {
  if (!isObject(value)) return null;
  return decodeWireRunReport(
    "grading_run_id" in value ? value : normalizedReportToWire(value),
  );
}

File: templates/langchain/rubric/{{cookiecutter.project_slug}}/src/{{cookiecutter.project_slug}}/contracts.py
from __future__ import annotations

import hashlib
import re
from collections.abc import Mapping, Sequence
from typing import Literal, NotRequired, TypedDict, cast

RunMode = Literal["demo", "live"]
RubricResult = Literal[
    "satisfied",
    "needs_revision",
    "max_iterations_reached",
    "failed",
    "grader_error",
]
TerminalStatus = Literal[
    "satisfied",
    "max_iterations_reached",
    "failed",
    "grader_error",
]
GateReason = Literal[
    "satisfied_with_current_evidence",
    "terminal_status_not_satisfied",
    "current_evidence_missing",
]
PublicErrorCode = Literal["invalid_input", "live_configuration", "runtime"]
UIEventType = Literal[
    "candidate",
    "grader_feedback",
    "rubric_evidence",
    "rubric_evaluation_start",
    "rubric_evaluation_end",
]

TASK_PROMPT = "Implement find_duplicates(values). Return valid Python source only."
BASELINE_RUBRIC = """\
- The response is valid Python source that defines find_duplicates(values).
- Each duplicated value appears exactly once in the returned list.
- Result order follows when each value first becomes a duplicate.
- Unhashable values such as nested lists are supported.
- The input sequence is not mutated.
- Before satisfied, call run_test_suite with the exact current candidate and receive ok=true.
"""
DEFAULT_MAX_ITERATIONS = 3
MAX_ITERATIONS = 20
MAX_CANDIDATE_CHARS = 3500
UI_EVENT_TYPES: tuple[UIEventType, ...] = (
    "candidate",
    "grader_feedback",
    "rubric_evidence",
    "rubric_evaluation_start",
    "rubric_evaluation_end",
)


class RunInput(TypedDict):
    rubric: str
    max_iterations: int


class EvidenceResult(TypedDict):
    candidate_id: str
    ok: bool
    behavior_failures: list[str]
    profile_failures: list[str]
    duration_ms: int
    timed_out: bool
    output_truncated: bool


class CandidateRecord(TypedDict):
    grading_run_id: str
    version: int
    iteration: int
    candidate_id: str
    source: str | None
    source_omitted: NotRequired[bool]


class EvidenceRecord(EvidenceResult):
    event_id: str
    grading_run_id: str
    iteration: int
    candidate_version: int
    requested_candidate_id: str


class CriterionResult(TypedDict):
    criterion: str
    passed: bool
    gap: str


class EvaluationEvent(TypedDict):
    event_id: str
    grading_run_id: str
    iteration: int
    candidate_version: int | None
    candidate_id: str | None
    result: RubricResult
    explanation: str
    criteria: list[CriterionResult]


class FeedbackRecord(TypedDict):
    event_id: str
    grading_run_id: str
    iteration: int
    candidate_version: int | None
    candidate_id: str | None
    message: str


class RunReport(TypedDict):
    mode: RunMode
    thread_id: str
    inner_thread_id: str
    grading_run_id: str
    terminal_status: TerminalStatus
    accepted: bool
    gate_reason: GateReason
    iterations: int
    candidates: list[CandidateRecord]
    final_candidate: str | None
    evidence: list[EvidenceRecord]
    evaluations: list[EvaluationEvent]
    feedback: list[FeedbackRecord]


class PublicError(TypedDict):
    code: PublicErrorCode
    message: str
    missing: list[str]


class UIEvent(TypedDict):
    event_id: str
    type: UIEventType
    grading_run_id: str
    iteration: int
    candidate_version: int | None
    candidate_id: str | None
    payload: dict[str, object]


_COMPLETE_PYTHON_FENCE = re.compile(
    r"\A[ \t]*```(?:python|py)[ \t]*\n(?P<source>.*)\n```[ \t]*\Z",
    flags=re.IGNORECASE | re.DOTALL,
)
_ALLOWED_INPUT_FIELDS = frozenset({"rubric", "max_iterations"})


class CandidateTooLongError(ValueError):
    """The normalized Worker output cannot cross the public/execution boundary."""


def _extract_text_content(value: object) -> str:
    if isinstance(value, str):
        return value
    if isinstance(value, Mapping):
        text = value.get("text")
        return text if isinstance(text, str) else ""
    if isinstance(value, Sequence) and not isinstance(value, (bytes, bytearray)):
        return "".join(_extract_text_content(item) for item in value)
    content = getattr(value, "content", None)
    if content is not None and content is not value:
        return _extract_text_content(content)
    raise TypeError("Candidate content must contain text.")


def normalize_candidate_source(value: object) -> str:
    """Return the one source representation used for hashing and execution."""
    source = _extract_text_content(value).replace("\r\n", "\n").replace("\r", "\n")
    if match := _COMPLETE_PYTHON_FENCE.fullmatch(source):
        source = match.group("source")
    return source.rstrip("\n") + "\n"


def candidate_id(value: object) -> str:
    source = normalize_candidate_source(value)
    return hashlib.sha256(source.encode("utf-8")).hexdigest()


def _require_positive_integer(value: object, name: str) -> int:
    if type(value) is not int or value < 1:
        raise ValueError(f"{name} must be a positive integer")
    return value


def _require_non_negative_integer(value: object, name: str) -> int:
    if type(value) is not int or value < 0:
        raise ValueError(f"{name} must be a non-negative integer")
    return value


def _require_candidate_source_limit(source: str) -> None:
    if len(source) > MAX_CANDIDATE_CHARS:
        raise CandidateTooLongError(f"normalized candidate source must be at most {MAX_CANDIDATE_CHARS} characters")


def build_candidate_record(
    *,
    grading_run_id: str,
    version: int,
    iteration: int,
    source: object,
) -> CandidateRecord:
    version = _require_positive_integer(version, "candidate version")
    iteration = _require_non_negative_integer(iteration, "candidate iteration")
    normalized = normalize_candidate_source(source)
    _require_candidate_source_limit(normalized)
    return {
        "grading_run_id": grading_run_id,
        "version": version,
        "iteration": iteration,
        "candidate_id": candidate_id(normalized),
        "source": normalized,
    }


def build_rejected_candidate_record(
    *,
    grading_run_id: str,
    version: int,
    iteration: int,
    source: object,
) -> CandidateRecord:
    """Retain only identity for an oversized candidate, never an executable prefix."""
    version = _require_positive_integer(version, "candidate version")
    iteration = _require_non_negative_integer(iteration, "candidate iteration")
    normalized = normalize_candidate_source(source)
    if len(normalized) <= MAX_CANDIDATE_CHARS:
        raise ValueError("rejected candidate source must exceed the public source limit")
    return {
        "grading_run_id": grading_run_id,
        "version": version,
        "iteration": iteration,
        "candidate_id": candidate_id(normalized),
        "source": None,
        "source_omitted": True,
    }


def _validate_candidate_history(
    grading_run_id: str,
    candidates: Sequence[CandidateRecord],
) -> None:
    previous_version = 0
    for candidate in candidates:
        if candidate["grading_run_id"] != grading_run_id:
            raise ValueError("candidate history must belong to the same grading run")
        version = candidate["version"]
        if type(version) is not int or version <= previous_version:
            raise ValueError("candidate versions must be strictly increasing positive integers")
        _require_non_negative_integer(candidate["iteration"], "candidate iteration")
        if candidate.get("source_omitted") is True:
            if candidate["source"] is not None or re.fullmatch(r"[0-9a-f]{64}", candidate["candidate_id"]) is None:
                raise ValueError("omitted candidate source must retain only a valid candidate ID")
        else:
            normalized_source = normalize_candidate_source(candidate["source"])
            _require_candidate_source_limit(normalized_source)
            if candidate["candidate_id"] != candidate_id(normalized_source):
                raise ValueError("candidate ID must match normalized source")
        previous_version = version


def validate_run_input(value: Mapping[str, object]) -> RunInput:
    unknown = sorted(set(value) - _ALLOWED_INPUT_FIELDS)
    if unknown:
        raise ValueError(f"Unsupported input field: {unknown[0]}")

    rubric = value.get("rubric", BASELINE_RUBRIC)
    if not isinstance(rubric, str) or not rubric.strip():
        raise ValueError("rubric must be a non-empty string")

    max_iterations = value.get("max_iterations", DEFAULT_MAX_ITERATIONS)
    if isinstance(max_iterations, bool) or not isinstance(max_iterations, int) or max_iterations < 1:
        raise ValueError("max_iterations must be a positive integer")
    if max_iterations > MAX_ITERATIONS:
        raise ValueError(f"max_iterations must be at most {MAX_ITERATIONS}")

    return {"rubric": rubric, "max_iterations": max_iterations}


def make_public_error(
    code: PublicErrorCode,
    message: str,
    *,
    missing: Sequence[str] = (),
) -> PublicError:
    return {"code": code, "message": message, "missing": list(missing)}


def _has_current_passing_evidence(
    *,
    grading_run_id: str,
    final_candidate: str | None,
    candidates: Sequence[CandidateRecord],
    evidence: Sequence[EvidenceRecord],
) -> bool:
    if final_candidate is None or not candidates:
        return False

    current = candidates[-1]
    if current.get("source_omitted") is True or current["source"] is None:
        return False
    normalized_final = normalize_candidate_source(final_candidate)
    current_id = candidate_id(normalized_final)
    if (
        current["grading_run_id"] != grading_run_id
        or current["candidate_id"] != current_id
        or normalize_candidate_source(current["source"]) != normalized_final
    ):
        return False

    return any(
        record["grading_run_id"] == grading_run_id
        and record["iteration"] == current["iteration"]
        and record["candidate_version"] == current["version"]
        and record["candidate_id"] == current_id
        and record["requested_candidate_id"] == current_id
        and record["ok"] is True
        and record["timed_out"] is False
        and record["output_truncated"] is False
        for record in evidence
    )


def build_run_report(
    *,
    mode: RunMode,
    thread_id: str,
    inner_thread_id: str,
    grading_run_id: str,
    terminal_status: TerminalStatus,
    iterations: int,
    candidates: Sequence[CandidateRecord],
    final_candidate: str | None,
    evidence: Sequence[EvidenceRecord],
    evaluations: Sequence[EvaluationEvent],
    feedback: Sequence[FeedbackRecord],
) -> RunReport:
    _validate_candidate_history(grading_run_id, candidates)
    normalized_final = None if final_candidate is None else normalize_candidate_source(final_candidate)
    if terminal_status != "satisfied":
        accepted = False
        gate_reason: GateReason = "terminal_status_not_satisfied"
    elif _has_current_passing_evidence(
        grading_run_id=grading_run_id,
        final_candidate=normalized_final,
        candidates=candidates,
        evidence=evidence,
    ):
        accepted = True
        gate_reason = "satisfied_with_current_evidence"
    else:
        accepted = False
        gate_reason = "current_evidence_missing"

    return {
        "mode": mode,
        "thread_id": thread_id,
        "inner_thread_id": inner_thread_id,
        "grading_run_id": grading_run_id,
        "terminal_status": terminal_status,
        "accepted": accepted,
        "gate_reason": gate_reason,
        "iterations": iterations,
        "candidates": [cast(CandidateRecord, dict(item)) for item in candidates],
        "final_candidate": normalized_final,
        "evidence": [cast(EvidenceRecord, dict(item)) for item in evidence],
        "evaluations": [cast(EvaluationEvent, dict(item)) for item in evaluations],
        "feedback": [cast(FeedbackRecord, dict(item)) for item in feedback],
    }
